File: src/main/java/io/lavagna/model/CardLabelValue.java
/**
 * This file is part of lavagna.
 *
 * lavagna is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * lavagna is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with lavagna.  If not, see <http://www.gnu.org/licenses/>.
 */
package io.lavagna.model;

import io.lavagna.common.ConstructorAnnotationRowMapper.Column;
import io.lavagna.model.CardLabel.LabelType;

import java.util.Date;

import lombok.Getter;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

@Getter
public class CardLabelValue {

	private final int cardLabelValueId;
	private final int cardId;
	private final int labelId;
	private final Boolean useUniqueIndex;
	private final LabelType labelValueType;
	private final LabelValue value;

	public CardLabelValue(@Column("CARD_LABEL_VALUE_ID") int cardLabelValueId, @Column("CARD_ID_FK") int cardId,
			@Column("CARD_LABEL_ID_FK") int labelId,
			@Column("CARD_LABEL_VALUE_USE_UNIQUE_INDEX") Boolean useUniqueIndex,
			@Column("CARD_LABEL_VALUE_TYPE") LabelType labelValueType,
			@Column("CARD_LABEL_VALUE_STRING") String valueString,
			@Column("CARD_LABEL_VALUE_TIMESTAMP") Date valueDate, @Column("CARD_LABEL_VALUE_INT") Integer valueInt,
			@Column("CARD_LABEL_VALUE_CARD_FK") Integer valueCard,
			@Column("CARD_LABEL_VALUE_USER_FK") Integer valueUser,
			@Column("CARD_LABEL_VALUE_LIST_VALUE_FK") Integer valueList) {
		this.cardLabelValueId = cardLabelValueId;
		this.cardId = cardId;
		this.labelId = labelId;
		this.useUniqueIndex = useUniqueIndex;
		this.labelValueType = labelValueType;
		this.value = new LabelValue(valueString, valueDate, valueInt, valueCard, valueUser, valueList);
	}

	public CardLabelValue newValue(String value) {
		return new CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, LabelType.STRING, value, null,
				null, null, null, null);
	}

	public CardLabelValue newValue(Date date) {
		return new CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, LabelType.TIMESTAMP, null, date,
				null, null, null, null);
	}

	public CardLabelValue newValue(Integer integer) {
		return new CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, LabelType.INT, null, null,
				integer, null, null, null);
	}

	public CardLabelValue newCardValue(Integer cardId) {
		return new CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, LabelType.CARD, null, null, null,
				cardId, null, null);
	}

	public CardLabelValue newUserValue(Integer userId) {
		return new CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, LabelType.USER, null, null, null,
				null, userId, null);
	}

	public CardLabelValue newListValue(Integer listId) {
		return new CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, LabelType.CARD, null, null, null,
				null, null, listId);
	}

	public CardLabelValue newNullValue() {
		return new CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, LabelType.NULL, null, null, null,
				null, null, null);
	}

	public CardLabelValue newValue(LabelType type, LabelValue value) {
		return new CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, type, value.valueString,
				value.valueTimestamp, value.valueInt, value.valueCard, value.valueUser, value.valueList);
	}

	@Getter
	public static class LabelValue {
		private final String valueString;
		private final Date valueTimestamp;
		private final Integer valueInt;
		private final Integer valueCard;
		private final Integer valueUser;
		private final Integer valueList;

		public LabelValue(@Column("CARD_LABEL_VALUE_STRING") String valueString,
				@Column("CARD_LABEL_VALUE_TIMESTAMP") Date valueTimestamp,
				@Column("CARD_LABEL_VALUE_INT") Integer valueInt,
				@Column("CARD_LABEL_VALUE_CARD_FK") Integer valueCard,
				@Column("CARD_LABEL_VALUE_USER_FK") Integer valueUser,
				@Column("CARD_LABEL_VALUE_LIST_VALUE_FK") Integer valueList) {
			this.valueString = valueString;
			this.valueTimestamp = valueTimestamp;
			this.valueInt = valueInt;
			this.valueCard = valueCard;
			this.valueUser = valueUser;
			this.valueList = valueList;
		}

		public LabelValue(String valueString) {
			this(valueString, null, null, null, null, null);
		}

		public LabelValue(Date valueTimestamp) {
			this(null, valueTimestamp, null, null, null, null);
		}

		public LabelValue(Integer valueUser) {
			this(null, null, null, null, valueUser, null);
		}

		public LabelValue() {
			this(null, null, null, null, null, null);
		}

		@Override
		public boolean equals(Object obj) {
			if (obj == null || !(obj instanceof LabelValue)) {
				return false;
			}
			LabelValue lv = (LabelValue) obj;
			return new EqualsBuilder().append(valueString, lv.valueString).append(valueTimestamp, lv.valueTimestamp)
					.append(valueInt, lv.valueInt).append(valueCard, lv.valueCard).append(valueUser, lv.valueUser)
					.append(valueList, lv.valueList).isEquals();
		}

		@Override
		public int hashCode() {
			return new HashCodeBuilder().append(valueString).append(valueTimestamp).append(valueInt).append(valueCard)
					.append(valueUser).append(valueList).toHashCode();
		}
	}
}

File: src/main/kotlin/io/lavagna/model/CardLabelValue.kt
/**
 * This file is part of lavagna.

 * lavagna is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * lavagna is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with lavagna.  If not, see //www.gnu.org/licenses/>.
 */
package io.lavagna.model

import ch.digitalfondue.npjt.ConstructorAnnotationRowMapper.Column
import io.lavagna.model.CardLabel.LabelType
import org.apache.commons.lang3.builder.EqualsBuilder
import org.apache.commons.lang3.builder.HashCodeBuilder
import java.util.*

open class CardLabelValue(@Column("CARD_LABEL_VALUE_ID") val cardLabelValueId: Int,
                          @Column("CARD_ID_FK") val cardId: Int,
                          @Column("CARD_LABEL_ID_FK") val labelId: Int,
                          @Column("CARD_LABEL_VALUE_USE_UNIQUE_INDEX") val useUniqueIndex: Boolean?,
                          @Column("CARD_LABEL_VALUE_TYPE") val labelValueType: LabelType,
                          @Column("CARD_LABEL_VALUE_STRING") valueString: String?,
                          @Column("CARD_LABEL_VALUE_TIMESTAMP") valueDate: Date?,
                          @Column("CARD_LABEL_VALUE_INT") valueInt: Int?,
                          @Column("CARD_LABEL_VALUE_CARD_FK") valueCard: Int?,
                          @Column("CARD_LABEL_VALUE_USER_FK") valueUser: Int?,
                          @Column("CARD_LABEL_VALUE_LIST_VALUE_FK") valueList: Int?) {
    val value: LabelValue

    init {
        this.value = LabelValue(valueString, valueDate, valueInt, valueCard, valueUser, valueList)
    }

    fun newValue(value: String): CardLabelValue {
        return CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, LabelType.STRING, value, null,
            null, null, null, null)
    }

    fun newValue(date: Date): CardLabelValue {
        return CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, LabelType.TIMESTAMP, null, date,
            null, null, null, null)
    }

    fun newValue(integer: Int?): CardLabelValue {
        return CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, LabelType.INT, null, null,
            integer, null, null, null)
    }

    fun newCardValue(cardId: Int?): CardLabelValue {
        return CardLabelValue(cardLabelValueId, cardId!!, labelId, useUniqueIndex, LabelType.CARD, null, null, null,
            cardId, null, null)
    }

    fun newUserValue(userId: Int?): CardLabelValue {
        return CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, LabelType.USER, null, null, null,
            null, userId, null)
    }

    fun newListValue(listId: Int?): CardLabelValue {
        return CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, LabelType.CARD, null, null, null,
            null, null, listId)
    }

    fun newNullValue(): CardLabelValue {
        return CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, LabelType.NULL, null, null, null,
            null, null, null)
    }

    fun newValue(type: LabelType, value: LabelValue): CardLabelValue {
        return CardLabelValue(cardLabelValueId, cardId, labelId, useUniqueIndex, type, value.valueString,
            value.valueTimestamp, value.valueInt, value.valueCard, value.valueUser, value.valueList)
    }

    class LabelValue @JvmOverloads constructor(@Column("CARD_LABEL_VALUE_STRING") val valueString: String? = null,
                                               @Column("CARD_LABEL_VALUE_TIMESTAMP") val valueTimestamp: Date? = null,
                                               @Column("CARD_LABEL_VALUE_INT") val valueInt: Int? = null,
                                               @Column("CARD_LABEL_VALUE_CARD_FK") val valueCard: Int? = null,
                                               @Column("CARD_LABEL_VALUE_USER_FK") val valueUser: Int? = null,
                                               @Column("CARD_LABEL_VALUE_LIST_VALUE_FK") val valueList: Int? = null) {

        constructor(valueTimestamp: Date) : this(null, valueTimestamp, null, null, null, null) {
        }

        constructor(valueUser: Int?) : this(null, null, null, null, valueUser, null) {
        }

        override fun equals(obj: Any?): Boolean {
            if (obj == null || obj !is LabelValue) {
                return false
            }
            return EqualsBuilder().append(valueString, obj.valueString).append(if (valueTimestamp == null) null else valueTimestamp.time / 1000, if (obj.valueTimestamp == null) null else obj.valueTimestamp.time / 1000)//go down to second...
                .append(valueInt, obj.valueInt).append(valueCard, obj.valueCard).append(valueUser, obj.valueUser).append(valueList, obj.valueList).isEquals
        }

        override fun hashCode(): Int {
            return HashCodeBuilder().append(valueString).append(valueTimestamp).append(valueInt).append(valueCard).append(valueUser).append(valueList).toHashCode()
        }
    }
}
